File: problem-solving-patterns/two-pointers/intermediate/3-sum/solution.py
def three_sum(numbers):
    # Sort the array to enable the two-pointer approach
    numbers.sort()
    # Initialize an empty list to store valid triplets
    sums = []

    # Outer loop: Iterate through each number as the first element of potential triplets
    # Stop at len(numbers)-2 to ensure at least 3 elements remain
    for i in range(len(numbers) - 2):
        # Initialize two pointers:
        # - `low` starts right after the current number
        # - `high` starts at the end of the array
        low = i + 1
        high = len(numbers) - 1

        # Skip duplicate values for `i` to avoid duplicate triplets
        # (Compare with previous element to catch duplicates)
        if i > 0 and numbers[i] == numbers[i - 1]:
            continue

        # Inner loop: Two-pointer technique to find complementary pairs
        while low < high:
            # Calculate the current sum of the three numbers
            current_sum = numbers[i] + numbers[low] + numbers[high]

            if current_sum < 0:
                # Sum is too small: move `low` right to increase the sum
                low += 1
            elif current_sum > 0:
                # Sum is too large: move `high` left to decrease the sum
                high -= 1
            else:
                # Found a valid triplet that sums to zero
                sums.append([numbers[i], numbers[low], numbers[high]])
                # Move both pointers inward to search for new combinations
                low += 1
                high -= 1

                # Skip duplicate values for `low` (compare with previous element)
                while low < high and numbers[low] == numbers[low - 1]:
                    low += 1
                # Skip duplicate values for `high` (compare with next element)
                while low < high and numbers[high] == numbers[high + 1]:
                    high -= 1

    # Return all unique triplets that sum to zero
    return sums

File: problem-solving-patterns/two-pointers/intermediate/3-sum/solution.js
function threeSum(numbers) {
  // Sort the array in ascending order to enable the two-pointer approach
  // Time Complexity: O(n log n)
  numbers = numbers.sort((a, b) => a - b);

  // This will store all unique triplets that sum to 0
  const sums = [];

  // Outer loop: iterate through each number as the first element of potential triplets
  // We stop at length - 2 because we need at least 3 numbers to form a triplet
  // Time Complexity: O(n) for this loop
  for (let i = 0; i < numbers.length - 2; i++) {
    // Initialize two pointers:
    let low = i + 1; // 'low' starts right after current number
    let high = numbers.length - 1; // 'high' starts at end of array

    // Skip duplicate values for the first element to avoid duplicate triplets
    // We compare with previous element (i-1) to catch duplicates
    // This optimization helps reduce unnecessary computations
    if (i > 0 && numbers[i] === numbers[i - 1]) {
      continue;
    }

    // Inner loop: two-pointer approach to find complementary pairs
    // Time Complexity: O(n) for this while loop (nested, but not quadratic due to pointer movement)
    while (low < high) {
      // Calculate current sum of the three numbers
      const sum = numbers[i] + numbers[low] + numbers[high];

      if (sum < 0) {
        // Sum is too small, move left pointer right to increase sum
        low++;
      } else if (sum > 0) {
        // Sum is too large, move right pointer left to decrease sum
        high--;
      } else {
        // Found a valid triplet that sums to 0
        sums.push([numbers[i], numbers[low], numbers[high]]);

        // Move both pointers inward to search for new combinations
        low++;
        high--;

        // Skip duplicate values for the low pointer
        // This ensures we don't process the same value multiple times
        while (low < high && numbers[low] === numbers[low - 1]) {
          low++;
        }

        // Skip duplicate values for the high pointer
        // Similar optimization as above for the right pointer
        while (low < high && numbers[high] === numbers[high + 1]) {
          high--;
        }
      }
    }
  }

  // Return all found triplets
  // Space Complexity: O(n) in worst case (when many triplets exist)
  return sums;
}

export { threeSum };
